container/shim/src/fetchers/lassie.js

```diff
@@ -1,3 +1,143 @@
+import http from "node:http";
+import https from "node:https";
+
+import { CarBlockIterator } from "@ipld/car";
+import LRU from "lru-cache";
+import write from "stream-write";
+
+import { LASSIE_ORIGIN } from "../config.js";
+import { streamCAR, validateCarBlock } from "../utils/car.js";
+import { proxyResponseHeaders, toUtf8 } from "../utils/http.js";
+import { debug as Debug } from "../utils/logging.js";
+
+const debug = Debug.extend("lassie");
+const LASSIE_TIMEOUT = 120_000;
+
+const agentOpts = {
+  keepAlive: true,
+};
+const httpsAgent = new https.Agent(agentOpts);
+const httpAgent = new http.Agent(agentOpts);
+
+const blockCache = new LRU({
+  maxSize: 1024 ** 3, // 1 GB
+  sizeCalculation: (block) => Buffer.byteLength(block),
+  allowStale: true,
+});
+
 export function respondFromLassie(req, res, { cid, format }) {
-  // todo
+  debug(`Fetch ${req.path} from Lassie`);
+
+  const isRawFormat = format === "raw";
+  const isInBlockCache = isRawFormat && !req.params.path && blockCache.has(cid);
+  if (isInBlockCache) {
+    const block = blockCache.get(cid);
+    return respondFromBlockCache(req, res, cid, block);
+  }
+
+  const lassieUrl = new URL(LASSIE_ORIGIN + toUtf8(req.path));
+  for (const [key, val] of Object.entries(req.query)) {
+    lassieUrl.searchParams.set(key, toUtf8(val));
+  }
+  lassieUrl.searchParams.set("format", "car");
+
+  const controller = new AbortController();
+  const timeout = setTimeout(() => controller.abort(), LASSIE_TIMEOUT);
+
+  const _http = lassieUrl.protocol === "https:" ? https : http;
+  const agent = lassieUrl.protocol === "https:" ? httpsAgent : httpAgent;
+
+  const lassieReq = _http
+    .get(
+      lassieUrl,
+      {
+        agent,
+        timeout: LASSIE_TIMEOUT,
+        signal: controller.signal,
+      },
+      async (fetchRes) => {
```

Comment:
If this `async` function fails, it crashes the process vs just making the req/res cycle fail. This is because `http.get()` accepts a callback, and doesn't have `async` function support built in.

Options are basically using `fetch` (or another lib with `async` support), `utils.promisify` and/or `events.once()`.

Reply:
Idk why the function is `async` in the first place, it doesn't await anything. Same for http.get call in `ipfs-gateway.js`. Must be leftover code that wasn't cleaned up. Also I know there's an eslint rule that prevents `async` without `await`, but I guess it's not enabled for this repo.